pkg/security/san_pinning.go

```diff
@@ -0,0 +1,183 @@
+
+	// Error templates.
+	errXFCCParse          = "failed to parse PEM block into x509 certificate: %w"
+	errDecodeURL          = "could not decode certificate URL format: %w"
+	errAnnotationsMissing = "KymaCR '%s' does not have annotation `%s`"
+)
+
+type RequestVerifier struct {
+	Client client.Client
+	log    logr.Logger
+}
+
+func NewRequestVerifier(client client.Client) RequestVerifier {
+	return RequestVerifier{
+		Client: client,
+		log:    ctrl.Log.WithName("listener"),
+	}
+}
+
+// Verify verifies the given request by fetching the KymaCR given in the request payload
+// and comparing the SAN(subject alternative name) of the certificate with the SKR-domain of the KymaCR.
+// If the request can be verified 'nil' will be returned.
+func (v RequestVerifier) Verify(request *http.Request) error {
+	certificate, err := v.getCertificateFromHeader(request)
+	if err != nil {
+		return err
+	}
+
+	domain, err := v.getDomain(request)
+	if err != nil {
+		return err
+	}
+
+	if containsURI(certificate.URIs, domain) ||
+		containsDNS(certificate.DNSNames, domain) ||
+		containsIP(certificate.IPAddresses, domain) {
+		v.log.V(debugLogLevel).Info("Received request verified")
+		return nil
+	}
+	return errNotVerified
+}
+
+// getCertificateFromHeader extracts the XFCC header and pareses it into a valid x509 certificate.
+func (v RequestVerifier) getCertificateFromHeader(r *http.Request) (*x509.Certificate, error) {
+	// Fetch XFCC-Header data
+	xfccValue, ok := r.Header[xfccHeader]
+	if !ok {
+		return nil, errHeaderMissing
+	}
+	xfccData := strings.Split(xfccValue[0], headerValueSeparator)
+
+	// Extract raw certificate
+	var cert string
+	for _, keyValuePair := range xfccData {
+		if strings.Contains(keyValuePair, certificateKey) {
+			cert = strings.Split(keyValuePair, keyValueSeparator)[1]
+			break
+		}
+	}
+	if cert == "" {
+		return nil, errEmptyCert
+	}
+
+	// Decode URL-format
+	decodedValue, err := url.QueryUnescape(cert)
+	if err != nil {
+		return nil, fmt.Errorf(errDecodeURL, err) //nolint:goerr113
+	}
+	decodedValue = strings.Trim(decodedValue, "\"")
+
+	// Decode PEM block and parse certificate
+	block, _ := pem.Decode([]byte(decodedValue))
+	if block == nil {
+		return nil, errPemDecode
+	}
+	certificate, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf(errXFCCParse, err) //nolint:goerr113
+	}
+	v.log.V(debugLogLevel).Info("X-Forwarded-Client-Certificate",
+		"certificate", certificate)
+
+	return certificate, nil
+}
+
+// getDomain fetches the KymaCR, mentioned in the requests body, and returns the value of the SKR-Domain annotation.
+func (v *RequestVerifier) getDomain(request *http.Request) (string, error) {
+	body, err := io.ReadAll(request.Body)
+	if err != nil {
+		return "", err
+	}
+	defer request.Body.Close()
+	request.Body = io.NopCloser(bytes.NewBuffer(body))
+
+	watcherEvent := &types.WatchEvent{}
+	err = json.Unmarshal(body, watcherEvent)
```

Comment:
The watcher is later also used in the listener that is correct! But at this point unfortunately the watcher Event was not parsed. 

But since I agree with your suggestion. I will open a PR for enhancing the listener pkg, then then update the dependency here, so we do not have to parse it again 💯

Reply:
https://github.com/kyma-project/runtime-watcher/pull/91

Reply:
@jeremyharisch The other PR is merged and this here is not adapted yet - correct?

Reply:
Uhh good catch! Let me adjust it!